vulnerabilities/package_managers.py

```diff
@@ -0,0 +1,301 @@
+#  the import process.
+
+
+class VersionAPI:
+    def __init__(self, cache: Mapping[str, Set[str]] = None):
+        self.cache = cache or {}
+
+    def get(self, package_name: str) -> Set[str]:
+        return self.cache.get(package_name, set())
+
+
+class LaunchpadVersionAPI(VersionAPI):
+    async def load_api(self, pkg_set):
+        async with ClientSession(raise_for_status=True) as session:
+            await asyncio.gather(
+                *[self.set_api(pkg, session) for pkg in pkg_set if pkg not in self.cache]
+            )
+
+    async def set_api(self, pkg, session):
+        if pkg in self.cache:
+            return
+        url = (
+            "https://api.launchpad.net/1.0/ubuntu/+archive/"
+            "primary?ws.op=getPublishedSources&"
+            "source_name={}&exact_match=true".format(pkg)
+        )
+        try:
+            all_versions = set()
+            while True:
+                response = await session.request(method="GET", url=url)
+                resp_json = await response.json()
+                if resp_json["entries"] == []:
+                    self.cache[pkg] = {}
+                    break
+                for release in resp_json["entries"]:
+                    all_versions.add(release["source_package_version"])
+                if resp_json.get("next_collection_link"):
+                    url = resp_json["next_collection_link"]
+                else:
+                    break
+            self.cache[pkg] = all_versions
+        except (ClientResponseError, asyncio.exceptions.TimeoutError):
+            self.cache[pkg] = {}
+
+
+class PypiVersionAPI(VersionAPI):
+    def get(self, package_name: str) -> Set[str]:
+        package_name = package_name.strip()
+
+        if package_name not in self.cache:
+            releases = set()
+            try:
+                with urlopen(f"https://pypi.org/pypi/{package_name}/json") as response:
+                    json_file = json.load(response)
+                    releases = set(json_file["releases"])
+            except HTTPError as e:
+                if e.code == 404:
+                    # PyPi does not have data about this package
+                    pass
+                else:
+                    raise
+
+            self.cache[package_name] = releases
+
+        return self.cache[package_name]
+
+
+class CratesVersionAPI(VersionAPI):
+    def get(self, package_name: str) -> Set[str]:
+        package_name = package_name.strip()
+
+        if package_name not in self.cache:
+            releases = set()
+
+            try:
+                with urlopen(f"https://crates.io/api/v1/crates/{package_name}") as response:
+                    response = json.load(response)
+                    for version_info in response["versions"]:
+                        releases.add(version_info["num"])
+            except HTTPError as e:
+                if e.code == 404:
+                    pass
```

Comment:
Do you really want to pass on a 404?

Reply:
I just moved this from https://github.com/nexB/vulnerablecode/blob/a15c726a83b515cc4566b79110901056279bd31e/vulnerabilities/importers/safety_db.py#L154 to separate file in this PR. Since I had implemented this importer's first edition, the idea was that given a package which does not exist in the pypi db, the endpoint would emit a 404. The pass just returns an empty set of versions for such packages. I don't see how this could be better handled